packages/tldraw/src/lib/shapes/image/ImageShapeUtil.tsx

```diff
@@ -63,9 +63,25 @@ export class ImageShapeUtil extends BaseBoxShapeUtil<TLImageShape> {
 		const prefersReducedMotion = usePrefersReducedMotion()
 		const [staticFrameSrc, setStaticFrameSrc] = useState('')
 		const [loadedSrc, setLoadedSrc] = useState('')
+		const [temporarySrc, setTemporarySrc] = useState('')
 		const isSelected = shape.id === this.editor.getOnlySelectedShapeId()
 		const { asset, url } = useAsset(shape.id, shape.props.assetId, shape.props.w)
 
+		useEffect(() => {
+			if (!asset) return
+			const temporaryAssetPreview = this.editor.getTemporaryAssetPreview(asset.id)
+			if (!asset.props.src && temporaryAssetPreview) {
+				setTemporarySrc(temporaryAssetPreview)
+			}
+		}, [asset])
+
+		useEffect(() => {
+			if (loadedSrc && temporarySrc) {
+				this.editor.deRegisterTemporaryAssetPreview(temporarySrc)
```

Comment:
Similar to the note above about the "deregister", the pattern I'd expect is 

```ts
const dispose = editor.setTemporaryAssetPreview(...)
try {
    return await doTheActualUpload()
} finally {
    dispose()
}
```

Reply:
I can't think of where else this would be called from. If I could, I'd suggest we follow the API pattern from the open menus (ie get / add / delete).

Reply:
Yeah, I hear that — totally get why you both would expect that.

It doesn't really work in this case though. The problem is this: the code that is doing the uploading and, in turn, the `setTemporaryAssetPreview` is only the beginning of the whole process. You still have:
- the network request to _fetch_ the newly uploaded asset.
- the decoding time of the image (albeit shorter but can be longer for higher res images).

So, the reason that `ImageShapeUtil` "controls" this process to deregister is because it's the one that is consuming this token if you will. If the upload disposes it, it will be too soon and you'll potentially have a flicker of a missing image before the new one comes in. 

Alternatively, there's some options we could explore to keep it contained in the editor:
- it could be something where the editor auto-disposes these URLs after, say, a minute after upload is done. That's still not technically _correct_ because the download might take longer if you're on a slow connection and then it will look broken.
- or, it could be something where it's an MRU cache and we only have a max of X amount of URLs in there. But not sure how much memory we want to stress out here and getting that right might be dicey.